### PR TITLE
show game board in all the game phases

### PR DESCRIPTION
**Pull request by kuronekochomusuke**
show game board in all the game phases

#2193

- convert the report display to be based off of the StatusBarPhaseDisplay to show the game board in all of the phases.   
- remove the report display from the main gui and use the mini report dialog to view the phase reports.
- add buttons to the report phases to show the mini report dialog and the player list dialog.   
- when entering the report phases maybe show the mini report dialog and player list dialog.

image

### CODE REVIEW

**Comment by HammerGS** [comment]
Just starting to play around and look at this. I'll let you know anything I find, but very cool so far.

**Comment by HammerGS** [comment]
Played a longish game today. It the more I used it the more I found myself constantly moving windows around. Is this something that could be dockable?

**Comment by kuronekochomusuke** [comment]
which windows were an issue?

do you know if there is some library that implements docking panels?

**Comment by HammerGS** [comment]
Sorry for the delay. I found the Player list in particular always seemed to be in the way and I found I closed it everytime it appeared.  Only to have it reopen depending on the phase. I'm prone to changing the size of the main window depending on the turn which I think is part of the issue. 

I seem to recall testing this when mjog was developing it and the docking part worked in conjunction with  https://github.com/MegaMek/megamek/pull/3125https://github.com/MegaMek/megamek/pull/3125.  Which had great player feedback other than then bug of it constantly growing in size. 

I really love the report being available and for me, is a game changer in how the program is played. Which as led me to some thoughts on your report filter (which I'll open as a separate RFE). 

I feel like if it remembered positions and what was opened and closed it would be perfect, or the dream dock able.

**Comment by kuronekochomusuke** [comment]
you can turn off the player list  or report auto popup by unchecking them in the menu.   they currently auto popup on the reporting phases if checked.  the unit display and mini map auto popup on non-report phases if checked.

was thinking it maybe good to have a setting for the auto popups by phase, instead of using the check marks.   so you could set to show, hide, or manual for each phase for the unit display, mini map, report display, and the player list.     that way you could set them to display or hide how you liked.   the check mark system does not give you much flexibly in how it works.  

I think some of the unit display growing in  size may have been fixed in the tab and non-tabbed unit display pull request.   I was running it to some of that with it.   I fixed some things in the pilot and armor tab then.    

would yall like to try to make it dock with this pull request or in a new one?      was think to make make the report display and the unit display dockable, or just one to start to see if it works.  would try to do it like the Lounge -> Team Overview

**Comment by kuronekochomusuke** [comment]
added Auto Display setting to control when the Unit Display, Minimap, Round Report, and Player List are automatically shown.  Can set them to be shown, hidden, or controlled manually on Report Phases and Non-Report Phases.  

Allowing it to be set for every phase makes it too complicated, so i went with just  Report Phases and Non-Report Phases.  

image

**Comment by kuronekochomusuke** [comment]
- added "docking" for the unit display.  
- can "switch location" between window mode, left side of board and right side of board.
- will add mini report if this works fine.

image

still working thru stuff, but good enough to see how it works.

**Comment by kuronekochomusuke** [comment]
- added in "docking" for the mini report display
- can "switch location" between window mode, left side of board and right side of board.

image

**Comment by kuronekochomusuke** [comment]
- had to make the "docking" on left and right side of the board an advanced setting.  
- defaults to docking on the left
- Could not get it to remember the split pane positions correctly when one was docked on the left and the other docked on the right.

**Comment by kuronekochomusuke** [comment]
- add an advance setting to allow switching the between stacking multiple panels on the y axis or x axis
- defaults to y axis
- x axis setting is probably only usable if you have an wide screen monitor.

here is when it is set to x axis, the previous screenshot shows the y axis behavior.
image

**Comment by kuronekochomusuke** [comment]
this is ready for review now

**Comment by HammerGS** [comment]
I was playing with this morning and so far it's really cool. It struck me with the changes that I kept going to View to look for the round report. Would it make sense more the round report from the Game menu to the View menu? Maybe in the same area as the unit display?

**Comment by kuronekochomusuke** [comment]
can move it over there.  move the player list also?  it fits with those also

**Comment by HammerGS** [comment]
> can move it over there. move the player list also? it fits with those also

yeah, I think that makes sense, to me when you close it you want to "view" it again.

**Comment by kuronekochomusuke** [comment]
grouped the dialogs and overlays together in the view menu and moved them up some.

image

**Comment by HammerGS** [comment]
> grouped the dialogs and overlays together in the view menu and moved them up some.
> 
> image

I think that flow makes much more sense.

**Comment by kuronekochomusuke** [comment]
add a divider between dialogs and overlays

image

File: megamek/src/megamek/client/ui/dialogs/UnitDisplayDialog.java [hunk 1]
```diff
@@ -20,36 +20,75 @@
...
+
         this.clientGUI = clientGUI;
     }
     //endregion Constructors
 
-    //region Getters/Setters
-    public UnitDisplay getUnitDisplay() {
-        return unitDisplay;
+    private void savePref() {
+        if ((this.getSize().width * this.getSize().height) > 0) {
```
Comment (SJuliez) [inline megamek/src/megamek/client/ui/dialogs/UnitDisplayDialog.java:69]: It seems the "this." can be removed here

File: megamek/src/megamek/client/ui/dialogs/UnitDisplayDialog.java [hunk 2]
```diff
@@ -20,36 +20,75 @@
...
+        }
     }
 
-    public void setUnitDisplay(final UnitDisplay unitDisplay) {
-        this.unitDisplay = unitDisplay;
+    @Override
+    protected void processWindowEvent(WindowEvent e) {
+        super.processWindowEvent(e);
+        if ((e.getID() == WindowEvent.WINDOW_DEACTIVATED) || (e.getID() == WindowEvent.WINDOW_CLOSING)) {
+            savePref();
```
Comment (SJuliez) [inline megamek/src/megamek/client/ui/dialogs/UnitDisplayDialog.java:89]: As in the previous file: Could the savePref() call be moved up to the WindowClosing listener?

File: megamek/src/megamek/client/ui/dialogs/MiniReportDisplayDialog.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        GUIP.setMiniReportSizeWidth(getSize().width);
+        GUIP.setMiniReportSizeHeight(getSize().height);
+        GUIP.setMiniReportPosX(getLocation().x);
+        GUIP.setMiniReportPosY(getLocation().y);
+    }
+
+    @Override
+    protected void processWindowEvent(WindowEvent e) {
+        super.processWindowEvent(e);
+        if ((e.getID() == WindowEvent.WINDOW_DEACTIVATED) || (e.getID() == WindowEvent.WINDOW_CLOSING)) {
```
Comment (SJuliez) [inline megamek/src/megamek/client/ui/dialogs/MiniReportDisplayDialog.java:71]: Could the savePref() be moved up to the WindowClosing listener with the same result?

File: megamek/src/megamek/client/ui/swing/ClientGUI.java [hunk 4]
```diff
@@ -1430,13 +1478,69 @@ private void toggleAccessibilityWindow() {
...
         bv.refreshDisplayables();
     }
 
     /** Shows or hides the minimap based on the current menu setting. */
     private void maybeShowMinimap() {
-        setMapVisible(GUIP.getMinimapEnabled());
+        GamePhase phase = getClient().getGame().getPhase();
+
+        if (phase.isReport()) {
+            String action = GUIP.getMinimapAutoDisplayReportPhase();
```
Comment (SJuliez) [inline megamek/src/megamek/client/ui/swing/ClientGUI.java:1490]: This confuses me. Shouldn't a GUIP call return a boolean for something like this?

File: megamek/src/megamek/client/ui/swing/GUIPreferences.java [hunk 5]
```diff
@@ -189,14 +197,24 @@ public class GUIPreferences extends PreferenceStoreProxy {
...
-    public static final String MINIMAP_COLOURS = "MinimapColours";
-    public static final String MINIMAP_ENABLED = "MinimapEnabled";
-    public static final String MINIMAP_POS_X = "MinimapPosX";
-    public static final String MINIMAP_POS_Y = "MinimapPosY";
-    public static final String MINIMAP_ZOOM = "MinimapZoom";
+    public static final String MINI_REPORT_ENABLED = "MiniReportEnabled";
+    public static final String MINI_REPORT_AUTO_DISPLAY_REPORT_PHASE = "MiniReportAutoDiplayReportPhase";
+    public static final String MINI_REPORT_AUTO_DISPLAY_NONREPORT_PHASE = "MiniReportAutoDiplayNonReportPhase";
+    public static final String MINI_REPORT_LOCATION = "MiniReportLocation";
+    public static final String PLAYER_lIST_POS_X = "PlayerListPosX";
```
Comment (SJuliez) [inline megamek/src/megamek/client/ui/swing/GUIPreferences.java:204]: There is a non-capital L in the next few names

File: megamek/src/megamek/client/ui/swing/GUIPreferences.java [hunk 6]
```diff
@@ -485,6 +518,16 @@ protected GUIPreferences() {
         store.setDefault(MINI_REPORT_POS_Y, 150);
         store.setDefault(MINI_REPORT_SIZE_HEIGHT, 300);
         store.setDefault(MINI_REPORT_SIZE_WIDTH, 400);
+        store.setDefault(MINI_REPORT_ENABLED, true);
+        store.setDefault(MINI_REPORT_AUTO_DISPLAY_REPORT_PHASE, MSG_SHOW);
```
Comment (SJuliez) [inline megamek/src/megamek/client/ui/swing/GUIPreferences.java:522]: Repeated from above: This should be boolean values, unless there is a reason against that that Im not seeing. At the very least it shouldnt be localized strings.

File: megamek/src/megamek/client/ui/swing/PlayerListDialog.java [hunk 7]
```diff
@@ -19,42 +19,88 @@
...
 
     private Client client;
-
-    public PlayerListDialog(JFrame parent, Client client) {
-        super(parent, Messages.getString("PlayerListDialog.title"), false);
+    private JButton butOkay;
+    private boolean model;
+
+    private static final String MSG_OKAY = Messages.getString("Okay");
+    private static final String MSG_TITLE = Messages.getString("PlayerListDialog.title");
```
Comment (SJuliez) [inline megamek/src/megamek/client/ui/swing/PlayerListDialog.java:47]: You removed many of those static final localized strings in the other PR :) just saying, I don't mind

File: megamek/src/megamek/client/ui/swing/PlayerListDialog.java [hunk 8]
```diff
@@ -19,42 +19,88 @@
...
     private static final long serialVersionUID = 7270469195373150106L;
 
     private JList<String> playerList = new JList<>(new DefaultListModel<>());
 
     private Client client;
-
-    public PlayerListDialog(JFrame parent, Client client) {
-        super(parent, Messages.getString("PlayerListDialog.title"), false);
+    private JButton butOkay;
+    private boolean model;
```
Comment (SJuliez) [inline megamek/src/megamek/client/ui/swing/PlayerListDialog.java:44]: model?

**Review by SJuliez — CHANGES_REQUESTED** [review]
Tested it and liked it a lot. I have a few comments on the code.

File: megamek/src/megamek/client/ui/swing/PlayerListDialog.java (hunk 7 above)
Comment (SJuliez) [reply megamek/src/megamek/client/ui/swing/PlayerListDialog.java:47]: There is no need to change this, I just noticed it.

**Review by SJuliez — APPROVED** [review]

**Pull request merged**